### PR TITLE
Vectorize RandomlyZoomedCrop

### PR DESCRIPTION
**Pull request by james77777778**
# What does this PR do?

Related to #291

Benchmark results:

||Mode|n=1000|n=2000|n=5000|n=10000|
|-|-|-|-|-|-|
|OldRandomlyZoomedCrop|Eager|0.46s|0.61s|0.89s|1.16s|
|OldRandomlyZoomedCrop|Graph|0.05s|0.11s|0.41s|0.57s|
|RandomlyZoomedCrop|Eager|0.018s|0.020s|0.071s|0.158s|
|RandomlyZoomedCrop|Graph|0.007s|0.012s|0.061s|0.120s|

comparison

Both implementations fail to run with XLA due to lacking ImageProjectiveTransformV3 support.
https://github.com/tensorflow/tensorflow/issues/55194

### Visualization

<details>

from `examples/layers/preprocessing/classification/randomly_zoomed_crop.py`

```python
zoom_factor=(0.8, 1.2),
aspect_ratio_factor=(3.0 / 4.0, 4.0 / 3.0),
```
out1

```python
zoom_factor=(0.08, 2.0),
aspect_ratio_factor=(3.0 / 4.0, 4.0 / 3.0),
```
out2

</details>

### Some Notes

I have updated the logic to make this layer compatible for RaggedTensor. I want to share the idea behind it:

Original `get_random_transformation` needs `image` to get each shape, but it raises error when we simply migrate to `get_random_transformation_batch` with RaggedTensor inputs

https://github.com/keras-team/keras-cv/blob/fff847454802d343d3d8f5e6bf77296067d72283/keras_cv/layers/preprocessing/randomly_zoomed_crop.py#L88-L95

Without the information of `original_height` and `original_width`, the meanings of sampled transformations changed:
- new `new_heights` / `new_widths`: w/o normalization of `original_height` / `original_width`
- new `height_offsets` / `height_offsets`: just two uniform distributions from `0.0` to `1.0`

I compute the `transforms` in a new function `get_zoomed_crop_matrix` (which is similar to `get_zoom_matrix` in `RandomZoom`)

It is worth noting that `width_offsets` and `height_offsets` in `get_zoomed_crop_matrix` should be guaranteed to have identical distribution of original implementation

Take `height_offsets` as example:
original logic:
```python
        # get_random_transformation
        height_offset = self._random_generator.random_uniform(
            (),
            minval=tf.minimum(0.0, original_height - new_height),
            maxval=tf.maximum(0.0, original_height - new_height),
            dtype=tf.float32,
        )
        height_offset = height_offset / original_height
        
        # augment_image
        transform = RandomlyZoomedCrop._format_transform(
            [
                ...
                height_offset * height,
                ...
            ]
        )
```

For more detail:
```python
# original
_random_generator.random_uniform(
    (),
    minval=tf.minimum(0.0, original_height - new_height),
    maxval=tf.maximum(0.0, original_height - new_height),
    dtype=tf.float32,
)

# if (original_height - new_height) >= 0, it is equivalent to
_random_generator.random_uniform((), 0.0, original_height - new_height, tf.float32)
# should be equivalent to
_random_generator.random_uniform((), 0.0, 1.0, tf.float32) * (original_height - new_height)

# if (original_height - new_height) < 0, it is equivalent to
_random_generator.random_uniform((), original_height - new_height, 0.0, tf.float32)
# should be equivalent to
_random_generator.random_uniform((), 0.0, 1.0, tf.float32) * (original_height - new_height)

# so, it can be decuded to
_random_generator.random_uniform((), 0.0, 1.0, tf.float32) * (original_height - new_height)
```

new logic:
```python
        # get_random_transformation_batch
        height_offsets = self._random_generator.random_uniform(
            shape=(batch_size, 1), maxval=1.0, dtype=tf.float32
        )

        # get_zoomed_crop_matrix
        height_offsets = offsets[:, 1] * (image_heights - new_heights)
```

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/keras-team/keras-cv/blob/master/.github/CONTRIBUTING.md),
      Pull Request section?
- [X] Was this discussed/approved via a Github issue? Please add a link
      to it if that's the case.
- [X] Did you write any new necessary tests?
- [ ] If this adds a new model, can you run a few training steps on TPU in Colab to ensure that no XLA incompatible OP are used?

## Who can review?
@LukeWood @ianstenbit

### CODE REVIEW

**Review by LukeWood — CHANGES_REQUESTED** [review]
Wonderful PR!  Can you add a test like the one in https://github.com/keras-team/keras-cv/pull/1480/files#diff-1c771135874332a25b437da4a5706faea85ff719331dfe7beb255403a1a66477 as well?

**Comment by james77777778** [comment]
> Wonderful PR! Can you add a test like the one in https://github.com/keras-team/keras-cv/pull/1480/files#diff-1c771135874332a25b437da4a5706faea85ff719331dfe7beb255403a1a66477 as well?

Done.

In addition, I added `test_config_with_custom_name` and `test_output_dtypes` tests as well from RandomTranslation

**Comment by james77777778** [comment]
@ianstenbit 

I see the CI failure of keras-cv-accelerator-testing, but it should be unrelated to this PR?

**Comment by ianstenbit** [comment]
> I see the CI failure of keras-cv-accelerator-testing, but it should be unrelated to this PR?

@james77777778 yes this failure is unrelated to this PR. You can merge from master and re-push to get rid of the failure (or we can safely merge this as-is, but I'm going to wait for @LukeWood's review)

**Review by LukeWood — APPROVED** [review]
lgtm
